cp_diffdock_api.py

```diff
@@ -0,0 +1,261 @@
+        # get_id needs to run every time, so don't use default arguments
+        if baseDir is None:
+            baseDir = f"/tmp/diffdock{os.getpid()}"
+        if requestId is None:
+            requestId = requestIdGenerator.get_id()
+        work_dir = Path(baseDir) / f"dock_{requestId}"
+        return DiffDockOptions(
+            work_dir=work_dir
+        )
+
+
+class DiffDockNothingToDoException(Exception):
+    pass
+
+class DiffDockSetupException(Exception):
+    pass
+
+
+class DiffDockApi:
+    """
+    Wrapper for running DiffDock with DiffDockProtocol objects for requests and responses.
+    """
+    @staticmethod
+    def run_diffdock(request: DiffDockProtocol.Request, options: DiffDockOptions=None) -> DiffDockProtocol.Response:
+        try:
+            log.info("diffdock preparing request...")
+            preppedRequest = DiffDockApi.prepare_request(request, options)
+
+            args = InferenceArgs({
+                "protein_ligand_csv": preppedRequest.csv_file,
+                "out_dir": preppedRequest.out_dir,
+                "add_hs": request.add_hs,
+                "keep_hs": request.keep_hs,
+                "keep_src_3d": request.keep_src_3d,
+                "samples_per_complex": request.samples_per_complex,
+            })
+            log.info("diffdock running inference...")
+            RunInference(args)
+            response = DiffDockApi.process_results(preppedRequest)
+            log.info("diffdock prepared response.")
+            return response
+        except DiffDockNothingToDoException:
+            return DiffDockProtocol.Response.makeResults()
+        except Exception as ex:
+            return DiffDockProtocol.Response.makeError(f"DiffDock failed. {ex}")
+
+
+    @staticmethod
+    def prepare_request(docking_request: DiffDockProtocol.Request, options: DiffDockOptions=None) -> PreppedRequest:
+        """
+        DiffDock works on files in directories. This function prepares the pdb, sdf, and csv files needed
+        to fulfill the DiffDock request.
+        """
+
+        if len(docking_request.proteins) == 0 or len(docking_request.ligands) == 0:
+            raise DiffDockNothingToDoException()
+
+        # Create temp directory for this run
+        if options is None:
+            options = DiffDockOptions.make()
+
+        work_dir = options.work_dir
+        try:
+            os.makedirs(work_dir)
+        except OSError as e:
+            if e.errno != errno.EEXIST:
+                raise DiffDockSetupException(f"Could not prepare work directory, errcode {e.errno} ({e.strerror})")
+
+        # Write and remember pdb files
+        protein_entries: List[PreppedComplexUnit] = []
+        for protein_i, protein in enumerate(docking_request.proteins):
+            protein_name, proteinData = protein
+            protein_label = f"protein{protein_i}"
+            pdb_file = work_dir / f"{protein_label}.pdb"
+            try:
+                with open(pdb_file, 'w') as f:
```

Comment:
Consider explicit "wt" mode. As a long-time Mac and then Windows user, I used to hate Unix end-of-lines.